accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+  - STJ is considering removing its reference to using reflection emit, so referencing `System.Linq.Expressions.dll` would require keeping the the various System.Reflection.* references. Since reflection emit is currently implemented in the runtime (`System.Private.CoreLib`), disk savings by avoiding the API surface layer in `SREmit.ILGeneration` and `SREmit.Lightweight` is limited to ~30K.
+- Existing usages of STJ such as by the SDK and Visual Studio that do not use `dynamic` also do not want to carry `System.Linq.Expressions.dll`.
+
+The preferred approach is to ensure ILLinker can remove the dependency to `System.Linq.Expressions.dll` when `dynamic` is not needed. This is achieved by having an opt-in method `EnableDynamicTypes()` on `JsonSerializerOptions` that roots all internal usages of `System.Linq.Expressions.dll`.
+
+## Proposed API to enable dynamic mode
+All classes to be located in STJ.dll.
+
+```cs
+namespace System.Text.Json
+{
+    public partial class JsonSerializerOptions
+    {
+        // The root method that if not called, ILLinker will trim out the reference to `System.Linq.Expressions.dll`.
+        // STJ does not expose any types from SLE publically.
+        // Any System.Object-qualified type specified in a called to JsonSerializer.Deserialize() will create either a DynamicJsonArray, DynamicJsonObject or DynamicJsonValue.
+        public void EnableDynamicTypes();
+    }
+}
+```
+
+Sample code to enable dynamic types:
+```cs
+    var options = new JsonSerializerOptions();
+    options.EnableDynamicTypes();
+
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>("{\"MyProp\":42}", options);
+    int i = obj.MyProp;
+    Debug.Assert(i == 42);
+
+    string json = obj.Serialize();
+```
+
+### Varying the `T` in `Deserialize<T>`
+This
+```cs
+    dynamic obj = JsonSerializer.Deserialize<object>("{\"MyProp\":42}", options);
+```
+is equivalent to
+```cs
+    dynamic obj = JsonSerializer.Deserialize<dynamic>("{\"MyProp\":42}", options);
+```
+and these (assuming `options.EnableDynamicTypes()` was called)
+```cs
+    dynamic obj = JsonSerializer.Deserialize<DynamicJsonNode>("{\"MyProp\":42}", options);
+    dynamic obj = JsonSerializer.Deserialize<DynamicJsonObject>("{\"MyProp\":42}", options);
+```
+
+### Supporting non-dynamic to dynamic
+If a previously created non-dynamic `JsonNode` object needs dynamic support:
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    dynamic obj = node.ToDynamic();
+```
+
+The implementation of `ToDynamic()` is efficient since it will perform a shallow clone by copying the internal value (for `JsonValue`), array (for `JsonArray`) or dictionary (for `JsonObject`):
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    byte[] utf8 = node.SerializeToUtf8Bytes();
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>(utf8, options);
+```
+
+# Writable DOM
```

Comment:
Can you please share a code sample of how we expect to create a JSON DOM from scratch (object initialization/etc.)? That was considered an important scenario to make user-friendly as part of the first prototype.

Reply:
Here's one recently added sample that uses implicit operators + constructor initializers + array params:
```cs
var jObject = new JsonObject()
{
    ["MyString"] = "Hello!",
    ["MyNull"] = null,
    ["MyBoolean"] = false,
    ["MyInt"] = 43,
    ["MyDateTime"] = new DateTime(2020, 7, 8),
    ["MyGuid"] = new Guid("ed957609-cdfe-412f-88c1-02daca1b4f51"),
    ["MyArray"] = new JsonArray(2, 3, 42),
    ["MyObject"] = new JsonObject()
    {
        ["MyString"] = "Hello!!"
    },
    ["Child"] = new JsonObject()
    {
        ["ChildProp"] = 1
    }
};
```